src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -132,11 +142,20 @@ public void Dispose()
             _logger.LogTrace("Beginning receive loop");
             try
             {
-                while (!cancellationToken.IsCancellationRequested)
+                while (await _connection.Input.WaitToReadAsync(cancellationToken))
                 {
-                    // This is a little odd... we want to remove the InvocationRequest once and only once so we pull it out in the callback,
-                    // and stash it here because we know the callback will have finished before the end of the await.
-                    var message = await _adapter.ReadMessageAsync(_stream, _binder, cancellationToken);
+                    Message incomingMessage;
+                    if (!_connection.Input.TryRead(out incomingMessage))
```

Comment:
Why? No one else is reading so it should always succeed. If it fails (what would be the reason?) is there a guarantee that a subsequent call will succeed because if there isn't then we are in tight infinite loop that can't be completed (unless we also check the cancellation token inside the loop).